### PR TITLE
Add Azure eventhub support

### PR DESCRIPTION
**Pull request by bhapas**
Add support for Azure eventhub output.

For Credentials : https://github.com/Azure/azure-sdk-for-go/blob/6b6f76ebe0d2334c83e8b6f89af4fe9d0b1ce631/sdk/azidentity/README.md?plain=1#L156-L187

### Test Execution

```
workspace/stream - (azure-eventhub-output) >
 ./stream log main.go 
-p=azureeventhub 
--azure-event-hub-connection-string="Endpoint=sb://test-eventhub-stream-seis.servicebus.windows.net/;SharedAccessKeyName=RootManageSharedAccessKey;SharedAccessKey=SharedAccessKey"


{"level":"debug","ts":"2023-08-28T10:18:02.777+0200","caller":"output/util.go:28","msg":"Connecting...","address":""}
{"level":"info","ts":"2023-08-28T10:18:02.777+0200","caller":"output/util.go:41","msg":"Connected","address":""}
{"level":"debug","ts":"2023-08-28T10:18:02.777+0200","caller":"command/log.go:81","msg":"Sending log line.","address":"","log":"main.go","line_number":1}
{"level":"debug","ts":"2023-08-28T10:18:04.775+0200","caller":"command/log.go:81","msg":"Sending log line.","address":"","log":"main.go","line_number":2}
{"level":"debug","ts":"2023-08-28T10:18:05.101+0200","caller":"command/log.go:81","msg":"Sending log line.","address":"","log":"main.go","line_number":3}
{"level":"debug","ts":"2023-08-28T10:18:05.422+0200","caller":"command/log.go:81","msg":"Sending log line.","address":"","log":"main.go","line_number":4}
{"level":"debug","ts":"2023-08-28T10:18:05.560+0200","caller":"command/log.go:81","msg":"Sending log line.","address":"","log":"main.go","line_number":5}
{"level":"debug","ts":"2023-08-28T10:18:05.699+0200","caller":"command/log.go:81","msg":"Sending log line.","address":"","log":"main.go","line_number":6}
{"level":"debug","ts":"2023-08-28T10:18:05.831+0200","caller":"command/log.go:81","msg":"Sending log line.","address":"","log":"main.go","line_number":7}
{"level":"debug","ts":"2023-08-28T10:18:05.969+0200","caller":"command/log.go:81","msg":"Sending log line.","address":"","log":"main.go","line_number":8}
{"level":"debug","ts":"2023-08-28T10:18:06.108+0200","caller":"command/log.go:81","msg":"Sending log line.","address":"","log":"main.go","line_number":9}
{"level":"debug","ts":"2023-08-28T10:18:06.243+0200","caller":"command/log.go:81","msg":"Sending log line.","address":"","log":"main.go","line_number":10}
{"level":"debug","ts":"2023-08-28T10:18:06.385+0200","caller":"command/log.go:81","msg":"Sending log line.","address":"","log":"main.go","line_number":11}
{"level":"debug","ts":"2023-08-28T10:18:06.547+0200","caller":"command/log.go:81","msg":"Sending log line.","address":"","log":"main.go","line_number":12}
{"level":"debug","ts":"2023-08-28T10:18:06.689+0200","caller":"command/log.go:81","msg":"Sending log line.","address":"","log":"main.go","line_number":13}
{"level":"debug","ts":"2023-08-28T10:18:06.826+0200","caller":"command/log.go:81","msg":"Sending log line.","address":"","log":"main.go","line_number":14}
{"level":"debug","ts":"2023-08-28T10:18:07.028+0200","caller":"command/log.go:81","msg":"Sending log line.","address":"","log":"main.go","line_number":15}
{"level":"debug","ts":"2023-08-28T10:18:07.227+0200","caller":"command/log.go:81","msg":"Sending log line.","address":"","log":"main.go","line_number":16}
{"level":"debug","ts":"2023-08-28T10:18:07.381+0200","caller":"command/log.go:81","msg":"Sending log line.","address":"","log":"main.go","line_number":17}
{"level":"info","ts":"2023-08-28T10:18:07.516+0200","caller":"command/log.go:94","msg":"Log data sent.","address":"","log":"main.go","total_bytes":337,"total_lines":17}
workspace/stream - (azure-eventhub-output) > 
```

### Screenshots

Messages in Azure portal
image

### CODE REVIEW

**Review by P1llus — COMMENTED** [review]
At a quick glance, the main parts that I noticed is that since we don't emulate eventhub locally or in docker in any way, we need to add some more credential options.
Usually the Connection String or some sort of credentials file option would be great, and default back (with a log entry) when no credentials are found, that its going to use default or environment variable instead.

Check the golangci-lint output as well :)

**Comment by marc-gr** [comment]
It looks nice!

I wonder if we could add some tests for it if possible and to add some docs about how to use it.

File: pkg/output/lumberjack/lumberjack_test.go
Comment (efd6) [inline pkg/output/lumberjack/lumberjack_test.go:1]: Why is this changed?

**Review by efd6 — COMMENTED** [review]
I think this needs docs and probably at least one example configuration to help users understand how to hold it.

File: pkg/output/lumberjack/lumberjack_test.go
Comment (bhapas) [reply pkg/output/lumberjack/lumberjack_test.go:1]: Unrelated to this PR.. Will remove

File: pkg/output/azureeventhub/azure_event_hub.go [hunk 1]
```diff
@@ -0,0 +1,77 @@
...
+
+	return &Output{opts: opts, producerClient: producerClient}, nil
+}
+
+func (*Output) DialContext(_ context.Context) error {
+	return nil
+}
+
+func (o *Output) Close() error {
+	o.producerClient.Close(context.TODO())
```
Comment (andrewkroh) [inline pkg/output/azureeventhub/azure_event_hub.go:57]:
```suggestion
	o.producerClient.Close(context.Background())
```

File: pkg/output/azureeventhub/azure_event_hub.go [hunk 2]
```diff
@@ -0,0 +1,77 @@
...
+	return nil
+}
+
+func (o *Output) Write(b []byte) (int, error) {
+	batch, err := o.producerClient.NewEventDataBatch(context.TODO(), nil)
```
Comment (andrewkroh) [inline pkg/output/azureeventhub/azure_event_hub.go:62]: You could create a cancelable context in the constructor and store it in `Output`. Then pass that context into these calls such that when `Close()` is called that context is cancelled and this operations should stop.

Comment (andrewkroh) [reply pkg/output/azureeventhub/azure_event_hub.go:62]: Roughly. I don't think you want to modify `DialContext` to override the cancelCtx value.

And there is a https://pkg.go.dev/context#CancelFunc instead of writing `func()`. It gives clarity.

**Comment by bhapas** [comment]
Raised https://github.com/Azure/azure-sdk-for-go/issues/21758 for adding support for emulator in Azurite for Event Hub.

File: README.md [hunk 3]
```diff
@@ -128,3 +129,31 @@ The emulator does not require authentication.
 - `gcs-bucket`: The name of the GCS bucket that should be created, should not already exist.
 - `gcs-object`: The name of the GCS object that will be populated with the collected data, using the configured GCS bucket.
 - `gcs-projectid`: The related projectID used when creating the bucket, this is required to be changed from the default value when not using an emulator.
+
+## Azure Event Hub Output Reference
+
+The Azure Event Hub output is used to collect data from the azure event hub resource
+When specifying a (`--azure-event-hub-connection-string`) , it should be retrieved as mentioned [here](https://learn.microsoft.com/en-us/azure/event-hubs/event-hubs-get-connection-string).
```
Comment (efd6) [inline README.md:136]:
```suggestion
When specifying a `--azure-event-hub-connection-string`, it should be retrieved as described [here](https://learn.microsoft.com/en-us/azure/event-hubs/event-hubs-get-connection-string).
```

File: README.md [hunk 4]
```diff
@@ -128,3 +129,31 @@ The emulator does not require authentication.
...
+      - --retry=30
+      - -p=azureeventhub
+      - --azure-event-hub-connection-string="Endpoint=sb://test-eventhub-stream-seis.servicebus.windows.net/;SharedAccessKeyName=RootManageSharedAccessKey;SharedAccessKey=SharedAccessKey"
+      - /sample_logs/testdata.log
+```
+
+### Options
+
+- `azure-event-hub-connection-string`: The connection string to connect to the Event Hub
+- `azure-event-hub-namespace`: The FullyQualifiedNamespace is the Event Hubs namespace name (ex: myeventhub.servicebus.windows.net)
```
Comment (efd6) [inline README.md:158]: Using "FullyQualifiedNamespace" makes it look like a defined entity. This is true in the Go code, be we shouldn't expect the user to know this.

```suggestion
- `azure-event-hub-namespace`: The fully qualified domain name of the Event Hubs namespace. This it the Event Hubs namespace followed by `servicebus.windows.net` (e.g. myeventhub.servicebus.windows.net).
```

File: README.md [hunk 5]
```diff
@@ -128,3 +129,31 @@ The emulator does not require authentication.
...
+      - log
+      - --retry=30
+      - -p=azureeventhub
+      - --azure-event-hub-connection-string="Endpoint=sb://test-eventhub-stream-seis.servicebus.windows.net/;SharedAccessKeyName=RootManageSharedAccessKey;SharedAccessKey=SharedAccessKey"
+      - /sample_logs/testdata.log
+```
+
+### Options
+
+- `azure-event-hub-connection-string`: The connection string to connect to the Event Hub
```
Comment (efd6) [inline README.md:157]:
```suggestion
- `azure-event-hub-connection-string`: The connection string to connect to the Event Hub.
```

File: command/root.go [hunk 6]
```diff
@@ -80,11 +81,16 @@ func ExecuteContext(ctx context.Context) error {
...
 	rootCmd.PersistentFlags().BoolVar(&opts.GCPPubsubOptions.Clear, "gcppubsub-clear", true, "GCP Pubsub clear flag")
 
-	// GCS output flags.
+	// AzureBlobStorage output flags.
 	rootCmd.PersistentFlags().StringVar(&opts.AzureBlobStorageOptions.Container, "azure-blob-storage-container", "testcontainer", "Azure Blob Storage container name")
 	rootCmd.PersistentFlags().StringVar(&opts.AzureBlobStorageOptions.Blob, "azure-blob-storage-blob", "testblob", "Azure Blob Storage blob name")
 	rootCmd.PersistentFlags().StringVar(&opts.AzureBlobStorageOptions.Port, "azure-blob-storage-port", "10000", "HTTP port used to connect to the blob storage, used for emulators and CI")
 
+	// AzureEventHub output flags.
+	rootCmd.PersistentFlags().StringVar(&opts.AzureEventHubOptions.FullyQualifiedNamespace, "azure-event-hub-namespace", "myeventhub.servicebus.windows.net", "Azure Eventhub namespace")
```
Comment (efd6) [inline command/root.go:90]: nit: suggest attached comments on stanzas consistent with the other flag sets modulo the AzureBlogStorage flags: no blank lines between members of the set or the set's name.

File: pkg/output/azureeventhub/azure_event_hub.go [hunk 7]
```diff
@@ -0,0 +1,80 @@
...
+}
+
+func New(opts *output.Options) (output.Output, error) {
+	var producerClient *azeventhubs.ProducerClient
+	var err error
+
+	if opts.AzureEventHubOptions.ConnectionString != "" {
+		producerClient, err = azeventhubs.NewProducerClientFromConnectionString(opts.AzureEventHubOptions.ConnectionString, opts.AzureEventHubOptions.EventHubName, nil)
+		if err != nil {
+			return nil, fmt.Errorf("error while creating new eventhub producer client from connectionstring : %w", err)
```
Comment (efd6) [inline pkg/output/azureeventhub/azure_event_hub.go:34]:
```suggestion
			return nil, fmt.Errorf("error while creating new eventhub producer client from connection string: %w", err)
```

File: pkg/output/azureeventhub/azure_event_hub.go [hunk 8]
```diff
@@ -0,0 +1,80 @@
...
+		}
+	} else {
+		fmt.Print("no connectionstring was provided, falling back to default credentials or environment variable")
```
Comment (efd6) [inline pkg/output/azureeventhub/azure_event_hub.go:37]:
```suggestion
		fmt.Print("no connection string was provided, falling back to default credentials or environment variable")
```

File: pkg/output/azureeventhub/azure_event_hub.go [hunk 9]
```diff
@@ -0,0 +1,80 @@
...
+
+		// Credentials set as env variables - https://github.com/Azure/azure-sdk-for-go/blob/6b6f76ebe0d2334c83e8b6f89af4fe9d0b1ce631/sdk/azidentity/README.md?plain=1#L156-L187
+		defaultAzureCred, err := azidentity.NewDefaultAzureCredential(nil)
+		if err != nil {
+			return nil, fmt.Errorf("missing azure credentials in the environment variables : %w", err)
```
Comment (efd6) [inline pkg/output/azureeventhub/azure_event_hub.go:42]:
```suggestion
			return nil, fmt.Errorf("missing azure credentials in the environment variables: %w", err)
```

File: pkg/output/azureeventhub/azure_event_hub.go [hunk 10]
```diff
@@ -0,0 +1,80 @@
...
+
+	if opts.AzureEventHubOptions.ConnectionString != "" {
+		producerClient, err = azeventhubs.NewProducerClientFromConnectionString(opts.AzureEventHubOptions.ConnectionString, opts.AzureEventHubOptions.EventHubName, nil)
+		if err != nil {
+			return nil, fmt.Errorf("error while creating new eventhub producer client from connectionstring : %w", err)
+		}
+	} else {
+		fmt.Print("no connectionstring was provided, falling back to default credentials or environment variable")
+
+		// Credentials set as env variables - https://github.com/Azure/azure-sdk-for-go/blob/6b6f76ebe0d2334c83e8b6f89af4fe9d0b1ce631/sdk/azidentity/README.md?plain=1#L156-L187
```
Comment (efd6) [inline pkg/output/azureeventhub/azure_event_hub.go:39]:
```suggestion
		// Credentials set as env variables - https://github.com/Azure/azure-sdk-for-go/tree/main/sdk/azidentity#environment-variables
```
?

File: pkg/output/azureeventhub/azure_event_hub.go [hunk 11]
```diff
@@ -0,0 +1,80 @@
...
+		defaultAzureCred, err := azidentity.NewDefaultAzureCredential(nil)
+		if err != nil {
+			return nil, fmt.Errorf("missing azure credentials in the environment variables : %w", err)
+		}
+
+		producerClient, err = azeventhubs.NewProducerClient(opts.AzureEventHubOptions.FullyQualifiedNamespace, opts.AzureEventHubOptions.EventHubName, defaultAzureCred, nil)
+		if err != nil {
+			return nil, fmt.Errorf("error while creating new eventhub producer client : %w", err)
```
Comment (efd6) [inline pkg/output/azureeventhub/azure_event_hub.go:47]:
```suggestion
			return nil, fmt.Errorf("error while creating new eventhub producer client: %w", err)
```

File: pkg/output/azureeventhub/azure_event_hub.go [hunk 12]
```diff
@@ -0,0 +1,80 @@
...
+func (o *Output) Close() error {
+	o.producerClient.Close(o.cancelCtx)
+	o.cancelFunc()
+	return nil
+}
+
+func (o *Output) Write(b []byte) (int, error) {
+	batch, err := o.producerClient.NewEventDataBatch(o.cancelCtx, nil)
+	if err != nil {
+		return 0, fmt.Errorf("error while creating new event data batch : %w", err)
```
Comment (efd6) [inline pkg/output/azureeventhub/azure_event_hub.go:67]:
```suggestion
		return 0, fmt.Errorf("error while creating new event data batch: %w", err)
```

File: pkg/output/azureeventhub/azure_event_hub.go [hunk 13]
```diff
@@ -0,0 +1,80 @@
...
+	}
+	eventData := azeventhubs.EventData{Body: b}
+
+	if err := batch.AddEventData(&eventData, nil); err != nil {
+		return 0, fmt.Errorf("error while adding data to event data batch : %w", err)
```
Comment (efd6) [inline pkg/output/azureeventhub/azure_event_hub.go:72]:
```suggestion
		return 0, fmt.Errorf("error while adding data to event data batch: %w", err)
```

File: pkg/output/azureeventhub/azure_event_hub.go [hunk 14]
```diff
@@ -0,0 +1,80 @@
...
+	}
+
+	if err := o.producerClient.SendEventDataBatch(context.TODO(), batch, nil); err != nil {
+		return 0, fmt.Errorf("error while sending event data batch : %w", err)
```
Comment (efd6) [inline pkg/output/azureeventhub/azure_event_hub.go:76]:
```suggestion
		return 0, fmt.Errorf("error while sending event data batch: %w", err)
```

File: pkg/output/lumberjack/lumberjack_test.go [hunk 15]
```diff
@@ -9,10 +9,11 @@ import (
 	"net"
 	"testing"
 
-	"github.com/elastic/go-lumber/server"
 	"github.com/stretchr/testify/assert"
 	"github.com/stretchr/testify/require"
 
+	"github.com/elastic/go-lumber/server"
+
```
Comment (efd6) [inline pkg/output/lumberjack/lumberjack_test.go:16]:
```suggestion
```

File: pkg/output/azureeventhub/azure_event_hub.go [hunk 16]
```diff
@@ -0,0 +1,80 @@
...
+	ctx, cancel := context.WithCancel(context.Background())
+	return &Output{opts: opts, producerClient: producerClient, cancelFunc: cancel, cancelCtx: ctx}, nil
+}
+
+func (*Output) DialContext(_ context.Context) error {
+	return nil
+}
+
+func (o *Output) Close() error {
+	o.producerClient.Close(o.cancelCtx)
```
Comment (efd6) [inline pkg/output/azureeventhub/azure_event_hub.go:59]: The really do this? That's awful; they don't even use the context in the close path.

File: pkg/output/azureeventhub/azure_event_hub.go [hunk 17]
```diff
@@ -0,0 +1,80 @@
...
+package azureeventhub
+
+import (
+	"context"
+	"fmt"
+
+	"github.com/elastic/stream/pkg/output"
+
+	"github.com/Azure/azure-sdk-for-go/sdk/azidentity"
+	"github.com/Azure/azure-sdk-for-go/sdk/messaging/azeventhubs"
```
Comment (efd6) [inline pkg/output/azureeventhub/azure_event_hub.go:13]:
```suggestion
	"github.com/Azure/azure-sdk-for-go/sdk/azidentity"
	"github.com/Azure/azure-sdk-for-go/sdk/messaging/azeventhubs"

	"github.com/elastic/stream/pkg/output"
```

File: command/root.go (hunk 6 above)
Comment (bhapas) [reply command/root.go:90]: You mean `// Azure BlobStorage output falgs.` ?

There is no blank line within the set. It looks like this

```
	// Azure EventHub output flags.
	rootCmd.PersistentFlags().StringVar(&opts.AzureEventHubOptions.FullyQualifiedNamespace, "azure-event-hub-namespace", "myeventhub.servicebus.windows.net", "Azure Eventhub namespace")
	rootCmd.PersistentFlags().StringVar(&opts.AzureEventHubOptions.EventHubName, "azure-event-hub-name", "test-eventhub-seis", "Azure Eventhub name")
	rootCmd.PersistentFlags().StringVar(&opts.AzureEventHubOptions.ConnectionString, "azure-event-hub-connection-string", "connectionstring", "Azure Eventhub connection string")
```

File: pkg/output/azureeventhub/azure_event_hub.go (hunk 16 above)
Comment (bhapas) [reply pkg/output/azureeventhub/azure_event_hub.go:59]: Ofcourse... Do you have any alternative suggestions against this?

Comment (efd6) [reply pkg/output/azureeventhub/azure_event_hub.go:59]: No, it we have to use it as it is provided.

**Review by efd6 — APPROVED** [review]
LGTM
